pkg/beaconfqdn/analyzer.go

```diff
@@ -0,0 +1,415 @@
+
+			query["$push"] = bson.M{
+				"dat": bson.M{
+					"icfqdn": fqdn,
+					"icert":  1,
+					"cid":    a.chunk,
+				}}
+
+			// create selector for output
+			output.query = query
+			output.selector = src.BSONKey()
+
+		} else {
+
+			query["$set"] = bson.M{
+				"dat.$.icert": 1,
+				"dat.$.cid":   a.chunk,
+			}
+
+			// create selector for output
+			output.query = query
+			output.selector = hostSelector
+		}
+	}
+
+	return output
+}
+
+func (a *analyzer) hostBeaconQuery(score float64, src data.UniqueIP, fqdn string) updateInfo {
+	ssn := a.db.Session.Copy()
+	defer ssn.Close()
+
+	var output updateInfo
+
+	// create query
+	query := bson.M{}
+
+	// check if we need to update
+	// we do this before the other queries because otherwise if a beacon
+	// starts out with a high score which reduces over time, it will keep
+	// the incorrect high max for that specific destination.
+	var resListExactMatch []interface{}
+
+	maxBeaconMatchExactQuery := src.BSONKey()
+	maxBeaconMatchExactQuery["dat.mbfqdn"] = fqdn
+
+	_ = ssn.DB(a.db.GetSelectedDB()).C(a.conf.T.Structure.HostTable).Find(maxBeaconMatchExactQuery).All(&resListExactMatch)
+
+	// if we have exact matches, update to new score and return
+	if len(resListExactMatch) > 0 {
+		query["$set"] = bson.M{
+			"dat.$.max_beacon_fqdn_score": score,
+			"dat.$.mbfqdn":                fqdn,
+			"dat.$.cid":                   a.chunk,
+		}
+
+		// create selector for output
+		output.query = query
+
+		// using the same find query we created above will allow us to match and
+		// update the exact chunk we need to update
+		output.selector = maxBeaconMatchExactQuery
+
+		return output
+	}
+
+	// The below is only for cases where the ip is not currently listed as a max beacon
+	// for a source
+	// update max beacon score
+	newFlag := false
+	updateFlag := false
+
+	// this query will find any matching chunk that is reporting a lower
+	// max beacon score than the current one we are working with
+	maxBeaconMatchLowerQuery := src.BSONKey()
+	maxBeaconMatchLowerQuery["dat"] = bson.M{
+		"$elemMatch": bson.M{
+			"cid":                   a.chunk,
+			"max_beacon_fqdn_score": bson.M{"$lte": score},
+		},
+	}
+	// find matching lower chunks
+	var resListLower []interface{}
+
+	_ = ssn.DB(a.db.GetSelectedDB()).C(a.conf.T.Structure.HostTable).Find(maxBeaconMatchLowerQuery).All(&resListLower)
+
+	// if no matching chunks are found, we will set the new flag
+	if len(resListLower) <= 0 {
+
+		maxBeaconMatchUpperQuery := src.BSONKey()
+		maxBeaconMatchUpperQuery["dat"] = bson.M{
+			"$elemMatch": bson.M{
+				"cid":                   a.chunk,
+				"max_beacon_fqdn_score": bson.M{"$gte": score},
+			},
+		}
+
+		// find matching upper chunks
+		var resListUpper []interface{}
+		_ = ssn.DB(a.db.GetSelectedDB()).C(a.conf.T.Structure.HostTable).Find(maxBeaconMatchUpperQuery).All(&resListUpper)
+		// update if no upper chunks are found
+		if len(resListUpper) <= 0 {
+			newFlag = true
+		}
+	} else {
```

Comment:
The sets are updated according to the chunk values, so it would never overwrite a previous high score for another chunk. The logic is that if there are lower scores for a matching chunk (the one we are currently on), we just overwrite them. Otherwise, we push a new one in. And pushing it in is important, because it will push multiple values in if there are race conditions. Even though that sounds like it is undesirable, having two array elements in $dat with high beacon scores is what we want. The $max queries in all the display functions take care of picking the biggest number, and we do not have to worry about which value a multithreaded mongo will chose for the $max if we do a simple update query. I think we ran this hundreds of times and there was no way to guarantee correct results with multithreading unless we allowed it to push in multiple high values. It usually only results in one max beacon per host, but sometimes there are more and this does not hurt functionality since they have the same chunk id.